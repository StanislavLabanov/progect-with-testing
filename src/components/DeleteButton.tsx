import TrashIcon from "../icons/trash.png";

type Props = {
  onClick: () => void;
  disabled: boolean;
};
export const DeleteButton = ({ onClick, disabled }: Props) => {
  return (
    <button
      className="button button-with-icon"
      onClick={onClick}
      disabled={disabled}
      data-alt="Удалить"
      data-testid="delete_button"
    >
      <img src={TrashIcon} alt="Удалить" />
    </button>
  );
};
