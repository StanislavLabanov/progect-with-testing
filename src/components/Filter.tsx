import { Dispatch, SetStateAction } from "react";

export interface Props {
   setActive: Dispatch<SetStateAction<boolean>>;
}

export const Filter = ({ setActive }: Props) => {
   return (
      <div className="filter-wrapper" data-testid="filter_block">
         Фильтровать
         <input data-testid="checkbox_filter" type="checkbox" onChange={() => setActive(prev => !prev)} />
      </div>
   );
};
