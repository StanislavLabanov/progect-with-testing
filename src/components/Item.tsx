import { DeleteButton } from "./DeleteButton";

type Props = Task & {
  onDelete: (id: Task["id"]) => void;
  onToggle: (id: Task["id"]) => void;
};

export const Item = (props: Props) => {
  return (
    <li className="item-wrapper">
      <input
        type="checkbox"
        id={props.id}
        defaultChecked={props.done}
        onChange={() => props.onToggle(props.id)}
        data-testid="checkbox_item"
      />
      <label data-testid="label_item" htmlFor={props.id} onClick={() => props.onToggle(props.id)}>
        {props.done ? <s>{props.header}</s> : props.header}
      </label>
      <DeleteButton
        disabled={!props.done}
        onClick={() => props.onDelete(props.id)}
      />
    </li>
  );
};
