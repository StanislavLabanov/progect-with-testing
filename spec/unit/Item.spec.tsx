import { render, screen } from "@testing-library/react";
import { Item } from "src/components/Item";
import ue from "@testing-library/user-event";

const userEvent = ue.setup({ advanceTimers: jest.advanceTimersByTime });

describe('Элемент списка задач', () => {
    it('название не должно быть больше 32 символов', () => {
        const onDelete = jest.fn();
        const onToggle = jest.fn();

        render(<Item id={"1"} header={"купить молоко"} done={false} onToggle={onToggle} onDelete={onDelete} />)
        const label = screen.getByTestId('label_item')

        expect(label.innerHTML.length < 32).toBe(true)
    });
    it('название не должно быть пустым', () => {
        const onDelete = jest.fn();
        const onToggle = jest.fn();

        render(<Item id={"2"} header={"купить хлеб"} done={false} onToggle={onToggle} onDelete={onDelete} />)
        const label = screen.getByTestId('label_item')

        expect(label.innerHTML).not.toBe('')
    });
    it('нельзя удалять невыполненные задачи', async () => {
        const onDelete = jest.fn();
        const onToggle = jest.fn();

        render(<Item id={"3"} header={"купить сникерс"} done={true} onToggle={onToggle} onDelete={onDelete} />)
        const checkbox = screen.getByTestId('checkbox_item')
        const deleteButton = screen.getByTestId('delete_button')

        expect(checkbox).toBeChecked()
        await userEvent.click(deleteButton);
        expect(onDelete).toHaveBeenCalledTimes(1)
    })
    it('текст должен стать перечеркнутым после выполнения задачи', () => {
        const onDelete = jest.fn();
        const onToggle = jest.fn();

        const { getByTestId } = render(<Item id={"1"} header={"купить хлеб"} done={true} onToggle={onToggle} onDelete={onDelete} />)
        const checked = getByTestId('checkbox_item')
        const label = screen.getByTestId('label_item')

        expect(checked).toBeChecked()
        expect(label.innerHTML).toMatch(/<s>/i)
    })
    it('отображение задачи', () => {
        const onDelete = jest.fn();
        const onToggle = jest.fn();

        const { asFragment } = render(<Item id={"1"} header={"купить хлеб"} done={true} onToggle={onToggle} onDelete={onDelete} />)
        const firstRender = asFragment();

        expect(firstRender).toMatchSnapshot();
    })
});