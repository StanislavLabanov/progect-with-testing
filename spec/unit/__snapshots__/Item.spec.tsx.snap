// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Элемент списка задач отображение задачи 1`] = `
<DocumentFragment>
  <li
    class="item-wrapper"
  >
    <input
      checked=""
      data-testid="checkbox_item"
      id="1"
      type="checkbox"
    />
    <label
      data-testid="label_item"
      for="1"
    >
      <s>
        купить хлеб
      </s>
    </label>
    <button
      class="button button-with-icon"
      data-alt="Удалить"
      data-testid="delete_button"
    >
      <img
        alt="Удалить"
        src="[object Object]"
      />
    </button>
  </li>
</DocumentFragment>
`;
